src/typechecker.erl

```diff
@@ -2890,6 +2906,106 @@ check_clause(_Env, _ArgsTy, _ResTy, Term) ->
     io:format("DEBUG: check_clause term: ~p~n", [Term]),
     throw(check_clause).
 
+%% Refine types by matching clause. MatchedTys are the types exhausted by
+%% each pattern in the previous clause.
+refine_clause_arg_tys(Tys, MatchedTys, [], TEnv) ->
```

Comment:
Why do you insist on having no guards when refining a type? I assume you don't want to deal with refining types based on predicates like `is_integer/1` right now. But couldn't you just ignore these types of guards and refine the type in the pattern anyway?

Reply:
To be able to refine the types for the next clause, we need to know exactly why the previous clause didn't match. If we have any guard, we can't be sure that we are at the next clause because of the pattern mismatched, because it may be that the clause mismatched because of the guard alone.

```Erlang
-spec foo(a|b, boolean()) -> b.
foo(A, Guard) ->
    case A of
        a when Guard -> b;
        X            -> X  %% <--- may still be a|b, when Guard is false
    end.
```

Reply:
Ah, yes. Indeed.